### PR TITLE
Render window refactor

### PR DESCRIPTION
**Pull request by akaszynski**
Redo of https://github.com/pyvista/pyvista/pull/2964

I'm not entirely sure if the failures on `main` were due to #2964, but I want to be sure since this is a substantial change. I'll run the CI a bunch of times and check for stability.

### CODE REVIEW

File: pyvista/ext/plot_directive.py [hunk 1]
```diff
@@ -116,6 +116,7 @@
 # must enable BUILDING_GALLERY to to keep windows active
 # enable offscreen to hide figures when generating them
```
Comment (akaszynski) [inline pyvista/ext/plot_directive.py:117]:
```suggestion
# enable offscreen to hide figures when generating them.
```

**Comment by adeak** [comment]
In case you pushed the trivial commit to trigger CI: doesn't a close-reopen cycle trigger a CI run?

File: pyvista/ext/plot_directive.py [hunk 2]
```diff
@@ -114,8 +114,9 @@
 import jinja2  # Sphinx dependency.
 
 # must enable BUILDING_GALLERY to to keep windows active
```
Comment (adeak) [inline pyvista/ext/plot_directive.py:116]:
```suggestion
# must enable BUILDING_GALLERY to keep windows active
```

**Review by adeak — COMMENTED** [review]
Here's another one to commit though

**Comment by akaszynski** [comment]
> In case you pushed the trivial commit to trigger CI: doesn't a close-reopen cycle trigger a CI run?

Forgot about that trick. I'll try that provided there's history in the commit that's fine.

**Comment by akaszynski** [comment]
Seeing at least one bug:
```
Current thread 0x00007f2f662ec740 (most recent call first):
  Garbage-collecting
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/charts.py", line 4399 in deep_clean
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/renderer.py", line 1906 in clear
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/renderers.py", line 377 in clear
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/plotting.py", line 3591 in clear
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/plotting.py", line 3847 in close
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/plotting.py", line 5125 in __del__
```

**Comment by akaszynski** [comment]
Failure on Linux `(3.9, 9.1)`:
```
Current thread 0x00007ff1b4c30740 (most recent call first):
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/render_window.py", line 168 in render
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/render_window.py", line 208 in show
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/plotting.py", line 5877 in show
  File "/home/runner/work/pyvista/pyvista/tests/test_picking.py", line 132 in test_disable_picking
  File "/opt/hostedtoolcache/Python/3.9.13/x64/lib/python3.9/site-packages/_pytest/python.py", line 192 in pytest_pyfunc_call
```

**Comment by adeak** [comment]
> Failure on Linux `(3.9, 9.1)`:

I have that exact setup and I can't reproduce any issue locally. Does CI use OSMesa?

**Comment by akaszynski** [comment]
> Does CI use OSMesa?

`xvfb-run`. I'm testing locally and hammering the tests using [pytest-repeat](https://pypi.org/project/pytest-repeat/).

This is going to be a fun one to debug. In the chance that it's a false positive, I've testing on #3217, but given that @larsoner reported test failures, I'm guessing something's up.

**Comment by adeak** [comment]
I also couldn't reproduce with `xvfb-run`, but I take it neither can you, and you are hammering exactly for the heisenbug to reveal itself.

If we can't reproduce this locally and yet CI breaks here and downstream, we have to consider what the substantial difference is between CI and local.

**Comment by akaszynski** [comment]
Failure on Windows (3.9):
```
tests/test_composite.py::test_multiblock_ref PASSED                      [ 10%]
Windows fatal exception: access violation

Thread 0x00000ff4 (most recent call first):
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 316 in wait
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 581 in wait
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 1304 in run
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 980 in _bootstrap_inner
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 937 in _bootstrap

Thread 0x00000e78 (most recent call first):
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 316 in wait
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 581 in wait
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 1304 in run
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 980 in _bootstrap_inner
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 937 in _bootstrap

Thread 0x00001028 (most recent call first):
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 316 in wait
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 581 in wait
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 1304 in run
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 980 in _bootstrap_inner
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 937 in _bootstrap

Thread 0x0000133c (most recent call first):
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 316 in wait
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 581 in wait
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 1304 in run
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 980 in _bootstrap_inner
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 937 in _bootstrap

Thread 0x00000e84 (most recent call first):
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 316 in wait
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 581 in wait
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 1304 in run
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 980 in _bootstrap_inner
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 937 in _bootstrap

Thread 0x00001564 (most recent call first):
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 316 in wait
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 581 in wait
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 1304 in run
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 980 in _bootstrap_inner
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 937 in _bootstrap

Thread 0x00000cc0 (most recent call first):
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 316 in wait
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 581 in wait
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 1304 in run
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 980 in _bootstrap_inner
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 937 in _bootstrap

Thread 0x000018fc (most recent call first):
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 316 in wait
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 581 in wait
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 1304 in run
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 980 in _bootstrap_inner
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 937 in _bootstrap

Thread 0x0000078c (most recent call first):
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 316 in wait
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 581 in wait
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 1304 in run
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 980 in _bootstrap_inner
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\threading.py", line 937 in _bootstrap

Current thread 0x000019c4 (most recent call first):
  File "D:\a\pyvista\pyvista\pyvista\plotting\render_window.py", line 185 in finalize
  File "D:\a\pyvista\pyvista\pyvista\plotting\plotting.py", line 3818 in _clear_ren_win
  File "D:\a\pyvista\pyvista\pyvista\plotting\plotting.py", line 3848 in close
  File "D:\a\pyvista\pyvista\pyvista\plotting\plotting.py", line 5125 in __del__
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\vtkmodules\util\numpy_support.py", line 226 in vtk_to_numpy
  File "D:\a\pyvista\pyvista\pyvista\utilities\helpers.py", line 190 in convert_array
  File "D:\a\pyvista\pyvista\pyvista\core\pyvista_ndarray.py", line 24 in __new__
  File "D:\a\pyvista\pyvista\pyvista\core\datasetattributes.py", line 543 in get_array
  File "D:\a\pyvista\pyvista\pyvista\utilities\helpers.py", line 222 in _assoc_array
  File "D:\a\pyvista\pyvista\pyvista\utilities\helpers.py", line 248 in point_array
  File "D:\a\pyvista\pyvista\pyvista\utilities\helpers.py", line 443 in get_array_association
  File "D:\a\pyvista\pyvista\pyvista\core\dataset.py", line 659 in set_active_scalars
  File "D:\a\pyvista\pyvista\pyvista\core\composite.py", line 971 in set_active_scalars
  File "D:\a\pyvista\pyvista\tests\test_composite.py", line 641 in test_set_active_scalars
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\python.py", line 192 in pytest_pyfunc_call
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\pluggy\_callers.py", line 39 in _multicall
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\pluggy\_manager.py", line 80 in _hookexec
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\pluggy\_hooks.py", line 265 in __call__
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\python.py", line 1761 in runtest
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\runner.py", line 166 in pytest_runtest_call
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\pluggy\_callers.py", line 39 in _multicall
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\pluggy\_manager.py", line 80 in _hookexec
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\pluggy\_hooks.py", line 265 in __call__
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\runner.py", line 259 in <lambda>
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\runner.py", line 338 in from_call
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\runner.py", line 258 in call_runtest_hook
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\runner.py", line 219 in call_and_report
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\runner.py", line 130 in runtestprotocol
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\runner.py", line 111 in pytest_runtest_protocol
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\pluggy\_callers.py", line 39 in _multicall
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\pluggy\_manager.py", line 80 in _hookexec
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\pluggy\_hooks.py", line 265 in __call__
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\main.py", line 347 in pytest_runtestloop
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\pluggy\_callers.py", line 39 in _multicall
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\pluggy\_manager.py", line 80 in _hookexec
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\pluggy\_hooks.py", line 265 in __call__
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\main.py", line 322 in _main
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\main.py", line 268 in wrap_session
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\main.py", line 315 in pytest_cmdline_main
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\pluggy\_callers.py", line 39 in _multicall
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\pluggy\_manager.py", line 80 in _hookexec
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\pluggy\_hooks.py", line 265 in __call__
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\config\__init__.py", line 164 in main
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\_pytest\config\__init__.py", line 187 in console_main
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\pytest\__main__.py", line 5 in <module>
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\runpy.py", line 87 in _run_code
  File "C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\runpy.py", line 197 in _run_module_as_main
tests/test_composite.py::test_set_active_scalars 
```

**Comment by akaszynski** [comment]
Failures on Linux 3.10

```
tests/test_datasetattributes.py::test_set_array_string_lists_should_equal Fatal Python error: Segmentation fault

Thread 0x00007f2b358b6700 (most recent call first):
  File "/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/psutil/_common.py", line 727 in open_binary
  File "/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/psutil/_pslinux.py", line 415 in virtual_memory
  File "/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/psutil/__init__.py", line 1968 in virtual_memory
  File "/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytest_memprof.py", line 39 in <lambda>
  File "/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/pytest_memprof.py", line 47 in report_mem
  File "/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/threading.py", line 1378 in run
  File "/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/threading.py", line 1016 in _bootstrap_inner
  File "/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/threading.py", line 973 in _bootstrap

Current thread 0x00007f2bba583740 (most recent call first):
  Garbage-collecting
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/charts.py", line 4399 in deep_clean
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/renderer.py", line 1906 in clear
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/renderers.py", line 377 in clear
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/plotting.py", line 3598 in clear
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/plotting.py", line 3854 in close
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/plotting.py", line 5132 in __del__
  File "/opt/hostedtoolcache/Python/3.10.6/x64/lib/python3.10/site-packages/hypothesis/internal/conjecture/data.py", line 563 in 
```

And Linux 3.9:

```
tests/test_renderer.py::test_camera_position Fatal Python error: Segmentation fault

Thread 0x00007fb9f38a8700 (most recent call first):
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/renderer.py", line 2823 in deep_clean
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/renderers.py", line 310 in deep_clean
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/plotting.py", line 3878 in deep_clean
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/plotting.py", line 5133 in __del__
```

**Comment by adeak** [comment]
Do these segfaults seem to have any pattern to you? I've been wondering: I've had issues in the past with running pyvista in shells that had too low of a `ulimit -Sv` setting. I would get all sorts of errors, usually revolving around memory allocation, but often just a segfault. Any chance that there are memory issues (or memory limit issues) in CI leading to all the heisenbugs?

Or could you confirm any of these locally?

**Comment by akaszynski** [comment]
> Do these segfaults seem to have any pattern to you?

Yes, I'm seeing errors on `__del__`. I'm attempting to remove any dependency on `deep_clean` and additional destructors. I'm guessing there's some garbage collection race condition and I'm trying to just let Python garbage collect on its own.

File: pyvista/plotting/renderers.py [hunk 3]
```diff
@@ -297,16 +306,17 @@ def set_active_renderer(self, index_row, index_column=None):
...
-                    renderer.deep_clean()
+        """Clear all renderers.
+
+        .. deprecated:: 0.37.0
+           Use :func:`Renderers.clear` instead.
+
+        """
+        # plan to convert to error 0.40.0, remove at 0.43.0
+        warnings.warn(
+            '`deep_clean` has been depreciated, Use `Renderers.clear()`', PyvistaDeprecationWarning
```
Comment (adeak) [inline pyvista/plotting/renderers.py:317]:
```suggestion
            '`deep_clean` has been deprecated, Use `Renderers.clear()`', PyvistaDeprecationWarning
```

File: pyvista/plotting/renderer.py [hunk 4]
```diff
@@ -2773,51 +2826,48 @@ def remove_environment_texture(self):
...
-        if hasattr(self, '_box_object'):
-            self.remove_bounding_box(render=render)
-        if self._shadow_pass is not None:
-            self.disable_shadows()
-        if self.__charts is not None:
-            self.__charts.deep_clean()
-            self.__charts = None
+        # plan to convert to error 0.40.0, remove at 0.43.0
+        warnings.warn(
+            '`deep_clean` has been depreciated, Use `Renderer.clear()`', PyvistaDeprecationWarning
```
Comment (adeak) [inline pyvista/plotting/renderer.py:2859]:
```suggestion
            '`deep_clean` has been deprecated, Use `Renderer.clear()`', PyvistaDeprecationWarning
```

File: pyvista/plotting/plotting.py [hunk 5]
```diff
@@ -3878,16 +3883,17 @@ def close(self, render=False):
...
-        self.textActor = None
+        """Clean the plotter of the memory.
+
+        .. deprecated:: 0.37.0
+           Use :func:`Plotter.clear` instead.
+
+        """
+        # plan to convert to error 0.40.0, remove at 0.43.0
+        warnings.warn(
+            '`deep_clean` has been depreciated, Use `Plotter.clear()`', PyvistaDeprecationWarning
```
Comment (adeak) [inline pyvista/plotting/plotting.py:3894]:
```suggestion
            '`deep_clean` has been deprecated, Use `Plotter.clear()`', PyvistaDeprecationWarning
```

**Comment by akaszynski** [comment]
### Linux 3.10
```
tests/test_picking.py::test_enable_cell_picking_interactive Fatal Python error: Segmentation fault

Current thread 0x00007fa2ebfb7740 (most recent call first):
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/render_window.py", line 175 in render
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/render_window.py", line 218 in show
  File "/home/runner/work/pyvista/pyvista/pyvista/plotting/plotting.py", line 5914 in show
  File "/home/runner/work/pyvista/pyvista/tests/test_picking.py", line 160 in test_enable_cell_picking_interactive
```

**Comment by akaszynski** [comment]
Able to reproduce on a virtual machine on Azure with:

```
xvfb-run pytest --count=100
```

```
Current thread 0x00007f3004d85740 (most recent call first):                      File "/home/akaszyns/pyvista/pyvista/plotting/render_window.py", line 175 in 
render                                                                           File "/home/akaszyns/pyvista/pyvista/plotting/render_window.py", line 218 in 
show                                                                             File "/home/akaszyns/pyvista/pyvista/plotting/plotting.py", line 5914 in show
  File "/home/akaszyns/pyvista/tests/test_charts.py", line 1018 in test_iren_context_style                                                                    
  File "/home/akaszyns/pyvista/.venv/lib/python3.8/site-packages/_pytest/python.py", line 192 in pytest_pyfunc_call 
```

The render call bothers me.

**Comment by akaszynski** [comment]
Another one on the VM:
```
  File "/home/akaszyns/pyvista/pyvista/plotting/render_window.py", line 175 in render
File "/home/akaszyns/pyvista/pyvista/plotting/render_window.py", line 218 in show                                                                             File "/home/akaszyns/pyvista/pyvista/plotting/plotting.py", line 5914 in show
File "/home/akaszyns/pyvista/tests/test_picking.py", line 351 in test_enable_fly_to_right_click       
```

**Comment by akaszynski** [comment]
Ok. We're going with a debug build of VTK and if that fails I'm going to do this PR in micro-stages. This is driving me insane.

**Pull request closed without merging**